build.gradle.kts

```diff
@@ -5,3 +5,61 @@ subprojects {
         jcenter()
     }
 }
+
+val installVenv = tasks.register("installVenv", Exec::class.java) {
+    description = "Install a new virtualenv in ./venv"
+
+    outputs.dir("./venv")
+
+    commandLine("virtualenv", "venv")
+}
+
+val installPreCommit = tasks.register("installPreCommit", Exec::class.java) {
+    description = "Installs pre-commit in ./venv"
+
+    dependsOn(installVenv)
+
+    outputs.file("./venv/bin/pre-commit")
+
+    commandLine("./venv/bin/pip", "install", "pre-commit")
+}
+
+val installHooks = tasks.register("installHooks", Exec::class.java) {
+    description = "Run pre-commit hooks without installing them"
+
+    dependsOn(installPreCommit)
+
+    outputs.file(".git/hooks/pre-commit")
+    inputs.file(".pre-commit-config.yaml")
+
+    commandLine("./venv/bin/pre-commit", "install", "--install-hooks")
+}
+
+val runHooks = tasks.register("runHooks", Exec::class.java) {
+    description = "Run pre-commit hooks without installing them"
+
+    dependsOn(installPreCommit)
+
+    commandLine("./venv/bin/pre-commit", "run", "--all-files")
+}
+
+val preMerge = tasks.register("preMerge") {
+    description = "Runs pre-commit hooks, build the plugin and sample and execute tests."
+
+    dependsOn(runHooks)
+    dependsOn(gradle.includedBuild("gradle-plugin").task(":plugin:build"))
+    dependsOn(gradle.includedBuild("gradle-plugin").task(":plugin:check"))
+    dependsOn(subprojects.filter { it.name != "samples" }.map { it.tasks.named("assembleDebug") })
+    dependsOn(subprojects.filter { it.name != "samples" }.map { it.tasks.named("check") })
```

Comment:
This won't work at the moment, as you have to specify also the `mustRunAfter` rules. In the current situation you're just saying that `preMerge` depends on `runHooks`, `:plugin:build`, `:plugin:check`, `assembleDebug`, `check` but you're not specifying the order.

This is also the reason why Travis is failing (he's running a one of a sample's `check` before the `generateSwagger`).

Reply:
> This is also the reason why Travis is failing (he's running a one of a sample's check before the generateSwagger).

That was the point of this block int the root build.gradle.kts: 

```kotlin
        tasks.all {
            if (name == "assembleDebug") {
                dependsOn(tasks.named("generateSwagger"))
            }
        }
```

Except I forgot that the generated files are needed before the `assemble` task. I hopefully fixed that here: https://github.com/Yelp/swagger-gradle-codegen/pull/105/commits/b0e45e7a011b7ed44d9758907cd1888fbcd3f630#diff-392475fdf2bc320d17762ed97109a121R66